### ISSUE TITLE
A question about mask in 6th tutorial

### ISSUE THREAD

**Issue by hmtbgc** [bentrevett/pytorch-seq2seq#73]
Hi Ben, I met a question when reading your 6th tutorial. I feel puzzled about the **mask**.

In function "**make_src_mask**", "src_mask" has the size of **[batch size, 1, 1, src len]**.
However, in function "**make_trg_mask**", "trg_pad_mask" has the size of **[batch size, 1, trg len, 1]**.
They are both designed to mask the "pad". Why do they have different sizes?

I could persuade myself that "src_mask" has the size of **[batch size, 1, 1, src len]** since the 1st dimension could broadcast to **n_heads**, and the 2nd dimension could broadcast to **src_len**, which means the attention between kth word and all words has to be masked. But, how to explain "trg_pad_mask" has another size?

**Comment by M-e-r-c-u-r-y** [comment]
Hi hmtbc,

From my understanding, the target mask is designed to mask the padding words and the future words for any `i`th word. While decoding you wouldn't want to see the future target words before decoding the current target word as that would be like cheating/memorizing, that is why the matrix is `target len x target len` sized so that any `i`th word can only pay attention to itself and `i-1` words before it. The `trg_sub_mask` is used to achieve this. You can read more about this in the explanation provided in the Seq2Seq section.

Hope this helps :)

**Comment by hmtbgc** [comment]
Hi M-e-r-c-u-r-y, thanks for your response.

I could understand the principle of mask. However, I doubt why the size of `trg_pad_mask` should be  
` [batch size, 1, trg_len, 1]` rather than ` [batch size, 1,  1, trg_len]`, like the size of `src_mask`.

`trg_pad_mask` will be passed into `energy = energy.masked_fill(mask == 0, -1e10)`. The size of `energy` is `[batch_size, n_heads, trg_len, trg_len]`. Thus, the 0th and 1st dimension of `trg_pad_mask` could broadcast to `batch_size `and `n_heads`, but should the 2nd dimension broadcast to `trg_len` ?
Maybe the 3rd dimension should do this.


An example:
energy:
            I          can     swim   "pad"
I           0.70    0.15    0.15    0.00 
can      0.18    0.63    0.19    0.01
swim    0.15    0.17    0.66   0.01 
"pad"   0.02    0.14    0.04   0.80

Should trg_pad_mask's size be  [[1, 1, 1, 0]].repeat(4,1) or [[1],[1],[1],[0]].repeat(1,4) ?

**Comment by M-e-r-c-u-r-y** [comment]
The target pad mask's size in your example is 1,1,4,1
For this example, the batch size is 1, target length is 1. You've seem to have softmaxed the energies already which isn't the case for the energy, The rows needn't sum to 1 for the energies. However the concept can still be explained with these numbers.

 I  can  swim  "pad"
I  0.70  0.15  0.15  0.00
can  0.18  0.63  0.19  0.01
swim  0.15  0.17  0.66  0.01
"pad"  0.02  0.14  0.04  0.80

if you just use a target pad mask you would get

 I  can  swim  "pad"
I  0.70  0.15  0.15  0.00
can  0.18  0.63  0.19  0.00
swim  0.15  0.17  0.66  0.00
"pad"  0.00  0.00  0.00  0.00

but what you need is actually

 I  can  swim  "pad"
I  0.70  0.00 0.00  0.00
can  0.18  0.63  0.00 0.00
swim  0.15  0.17  0.66  0.00
"pad"  0.00  0.00  0.00  0.00

`I` can only pay attention to `I` and not to `can  swim  "pad"`
`can` can only pay attention to `I can` and not to  `swim  "pad"`
`swim` can only pay attention to `I can swim` and not to `"pad"`

To create this you need [`batch size,1,trg len, trg len`] . The `trg_pad_mask` of shape
[`batch size, 1, trg_len, 1`] interacts with `trg_sub_mask` of shape [`trg len, trg len`] to give [`batch size,1,trg len, trg len`] shaped mask.

Thus you need the mask to be of the shape [`batch size,1,trg len, trg len`] so that you broadcast the same over the various heads.

**Comment by hmtbgc** [comment]
OK, I got it. It is true that the original code is right.
Thanks for your help!
